### DISCUSSION TITLE
[RFC] Play a role in dependency resolution management

### DISCUSSION THREAD

**Discussion by d4mr** [vercel/turborepo#7490] [Ideas]
### What version of Turborepo are you using?

1.5.4

### What package manager are you using / does the bug impact?

pnpm

### What operating system are you using?

Mac

### Describe the Bug

I have a workspace with main app A and a library B. A had `react-query` as a dependency. B has `react-query` as a peerDependency. For library B to be able to work, the dependency needs to be a singleton.

Even though they are using the same exact version, the dependency inside node_modules is being symlinked to two different places:
In App A is being symlinked to: `node_modules/.pnpm/react-query@3.39.2_biqbaboplfbrettd7655fr4n2y/node_modules/react-query`
In Library B is being symlinked to: `node_modules/.pnpm/react-query@3.39.2/node_modules/react-query`

I have tried `pnpm up -r` and `syncpack`. I have also tried to set an override for react-query: "^3.39.2".
None of this solved my problem.

Additionally, I have tried `rm -rf`ing my `node_modules` and my `pnpm-lock.yaml`. Nothing worked.

Here are the two different listings for react-query in the pnpm lock file:
```
  /react-query/3.39.2:
    resolution: {integrity: sha512-F6hYDKyNgDQfQOuR1Rsp3VRzJnWHx6aRnnIZHMNGGgbL3SBgpZTDg8MQwmxOgpCAoqZJA+JSNCydF1xGJqKOCA==}
    peerDependencies:
      react: ^16.8.0 || ^17.0.0 || ^18.0.0
      react-dom: '*'
      react-native: '*'
    peerDependenciesMeta:
      react-dom:
        optional: true
      react-native:
        optional: true
    dependencies:
      '@babel/runtime': 7.19.0
      broadcast-channel: 3.7.0
      match-sorter: 6.3.1
    dev: true

  /react-query/3.39.2_biqbaboplfbrettd7655fr4n2y:
    resolution: {integrity: sha512-F6hYDKyNgDQfQOuR1Rsp3VRzJnWHx6aRnnIZHMNGGgbL3SBgpZTDg8MQwmxOgpCAoqZJA+JSNCydF1xGJqKOCA==}
    peerDependencies:
      react: ^16.8.0 || ^17.0.0 || ^18.0.0
      react-dom: '*'
      react-native: '*'
    peerDependenciesMeta:
      react-dom:
        optional: true
      react-native:
        optional: true
    dependencies:
      '@babel/runtime': 7.19.0
      broadcast-channel: 3.7.0
      match-sorter: 6.3.1
      react: 18.2.0
      react-dom: 18.2.0_react@18.2.0
    dev: false
``` 

The obvious difference seems to be dev with true and false for each of these. No clue why they both exist though. Some help here would be greatly appreciated, since our monorepo breaks if we cannot use the same version for these packages.

### Expected Behavior

The dependency should be reused, and the same dependency in both A and B should symlink to the same folder.

### To Reproduce

1. Create a new monorepo, create a new app with CRA. Install `react-query` in your this app.
2. Create a new package, with `react-query`, `react` and `react-dom` as a peer dependency.
3. Add the package as a dependency to the app.

Checking the symlinks as well as the pnpm-lock file shows that there are multiple installations of the `react-query` dependency.

**Comment by d4mr** [comment]
Here's the same issue I posted at `pnpm`: https://github.com/pnpm/pnpm/issues/5351
I have tried all the responses, and nothings works. A repro repo exists on that thread as well.

**Comment by chris-olszewski** [comment]
I need a little more clarity what turbo's role in this issue is. Could you elaborate on what behavior you expect turbo to have and also any turbo commands that lead to this behavior?

**Comment by IPWright83** [comment]
Having had a similar problem with a different package, I don't think is anything to do with Turborepo. You could remove Turborepo completely from the picture and it wouldn't make any difference.

**Comment by d4mr** [comment]
> I need a little more clarity what turbo's role in this issue is. Could you elaborate on what behavior you expect turbo to have and also any turbo commands that lead to this behavior?

@chris-olszewski while it looks to me like the issue is being caused by `pnpm`, I would expect for turborepo, being the monorepo I am using, to fix or deal with these problems.

`pnpm` is the recommended package manager for Turborepo, and sharing the same dependency between multiple apps/packages is a very common problem, so I would hope that there are measures/methods in place that allow me to achieve so using Turbo together with `pnpm`.

The entire reason we shifted to a monorepo was being able to share code easier, and so far, we have only been able to share basic presentational react components. Anything involving `react-context` requires a lot of debugging and manual adjustments of dependency versions to get everything playing together smoothly.

And in the case of `react-query`, no amount of tweaking makes this work. Mainly, I just want to understand whether this is simple and I am missing something, or this is something that `turborepo` cannot achieve right now.

**Comment by chris-olszewski** [comment]
@d4mr I agree this is something that's related to Turborepo and verifying all packages are using the same version of a dependency would be a good feature to have. I just wanted to make sure that this was more of a feature request and not a bug in the current behavior of Turborepo.
